File: spring-kt/src/main/kotlin/ovaphlow/crate/springkt/bulletin/Repository.kt
package ovaphlow.crate.springkt.bulletin

import org.springframework.data.annotation.Id
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Repository
import java.time.LocalDateTime

data class Bulletin(
    @Id val id: String,
    val title: String,
    val publishTime: LocalDateTime,
    val expireAt: LocalDateTime,
    val tag: String,
    val detail: String
)

@Repository
class Repository(
    private val jdbcTemplate: JdbcTemplate
) {
    fun filter(take: Int, skip: Long): List<Bulletin> {
        val q = """
            select cast(id as char) id, title, publish_time, expire_at, tag, detail
            from bulletin
            order by publish_time desc
            limit ?, ?
        """.trimIndent()
        val result = jdbcTemplate.query(q, arrayOf(skip, take)) { rs, _ ->
            Bulletin(
                rs.getString("id"),
                rs.getString("title"),
                rs.getTimestamp("publish_time").toLocalDateTime(),
                rs.getTimestamp("expire_at").toLocalDateTime(),
                rs.getString("tag"),
                rs.getString("detail")
            )
        }
        return result
    }

    fun filterByIdDetail(id: String, detail: String): Bulletin {
        val q = """
            select cast(id as char) id, title, publish_time, expire_at, tag, detail
            from bulletin
            where id = ? and json_contains(detail, ?)
        """.trimIndent()
        val result = jdbcTemplate.queryForObject(q, arrayOf(id, detail)) { rs, _ ->
            Bulletin(
                rs.getString("id"),
                rs.getString("title"),
                rs.getTimestamp("publish_time").toLocalDateTime(),
                rs.getTimestamp("expire_at").toLocalDateTime(),
                rs.getString("tag"),
                rs.getString("detail")
            )
        }
        return result!!
    }

    fun save(data: Bulletin) {
        val q = """
            insert into bulletin (
                id, title, publish_time, expire_at, tag, detail
            ) values (?, ?, ?, ?, ?, ?)
        """.trimIndent()
        jdbcTemplate.update(q, data.id, data.title, data.publishTime, data.expireAt, data.tag, data.detail)
    }
}

File: legacy/src/bulletin/repository.mjs
import { pool } from "../mysql.mjs";

const columns = ["id", "title", "published_at", "expired_at", "tag", "detail"];

export const filter = async (option, data) => {
  const client = pool.promise();
  if (option === "filterBy-id") {
    const q = `
        select ${columns.join(", ")}, cast(id as char) _id
        from crate.bulletin
        where id = ?
            and json_contains(detail, json_object("uuid", ?))
        `;
    const [result] = await client.query(q, [data["id"], data["uuid"]]);
    return result;
  }
  if (option === "filterBy-publishedAtRange-detail") {
    const q = `
        select ${columns.join(", ")}, cast(id as char) _id
        from crate.bulletin
        where published_at >= ?
            and published_at <= ?
            and json_contains(detail, ?)
        order by published_at desc
        limit ?, ?
        `;
    const [result] = await client.query(q, [
      data["time"],
      data["time1"],
      data["detail"],
      data["skip"],
      data["take"],
    ]);
    return result;
  }
};

export const save = async (data) => {
  const client = pool.promise();
  const q = `
    insert into crate.bulletin (
        ${columns.join(", ")}
    ) values (
        ?, ?, now(), ?, ?, ?
    )
    `;
  const [result] = await client.execute(q, [
    data["id"],
    data["title"],
    data["expiredAt"],
    data["tag"],
    data["detail"],
  ]);
  return result;
};
